README.org
* ol-mpv

org hyperlink for various video timestamps.
It provides link type that can specify time stamp of the video.
** Features
This only supports ~:follow~ functionality.

I've tried to implement ~:export~ before, but I'm ending up with giving
up as there's no way to access property drawer in export phase.

If you have good idea for support them, please open new issue and
let me know.
** Usage
*** 1. Set up URI in property

~ol-mpv~ uses ~OL_MPV_URI~ property with it.
Set it to URI that you want to associated with:

#+begin_src org
  ,* Some headline
  :PROPERTIES:
  :OL_MPV_URI: https://www.youtube.com/watch?v=mtbAUMzBWbE
  :END:
#+end_src

*** 2. Write timestamp as link
Then, you can write timestamp as a link.
You can write it just like you do in YouTube comment.

#+begin_src org
  mpv:01:00

  mpv:01:00
#+end_src

#+begin_src org
  [[mpv:01:00][Link at 1 minute]]
  [[mpv:01][Link at 1 second]]
  [[mpv:01:00:00][Link at 1 hour]]
  [[mpv:1:00][Link at 1 minute without first 0]]
  [[mpv:1][Link at 1 second without first 0]]
  [[mpv:1:00:00][Link at 1 hour without first 0]]
#+end_src

*** 3. Spawn mpv by open any link in there first time
When you open the link with help:org-open-at-point,
it will spawn mpv if it's not spawned yet.
*** 4. open link to change time position
When you open the link next time,
it will change time position.
** Supported video URIs
As this program uses mpv directly, all URIs supported by
mpv are available (not tested though).

For example:
+ https://
+ file:// (*this is discouraged, use normal path instead)
+ rtmp://
+ fd://


for list of all files, please execute the command below:

#+begin_src sh
  mpv --list-protocols | grep '://$' | sort | uniq
#+end_src

or visit https://mpv.io/manual/master/#protocols
** For debugging
If you have some problem with mpv, you might better
to investigate its output.

To do so, replace ~:buffer~ property of ~make-process~ with some buffer,
and remove ~"--no-terminal"~ Option from ~:command~.
(You can use string instead of buffer object, in which case it'll create buffer
with that name if it isn't exist yet.)

#+begin_src diff
    (let ((mpv-proc (make-process
		:name (format "ol-mpv mpv [%s]" video-uri)
  - 	      :buffer nil
  + 	      :buffer "BUFFER_NAME_YOU_WANT"
		:sentinel 'ol-mpv/mpv/sentinel
		:connection-type 'pipe
		:command `("mpv"
  - 			 "--no-terminal"
			   ,(format "--title=%s" (ol-mpv/mpv-WM-title video-uri))
			   "--no-input-terminal"
			   "--input-ipc-client=fd://0"
			   ,video-uri
			   ))))
#+end_src

This patch is for version 1.0.0, it might be changed
in the future.


CHANGELOG.org
* Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

** [Unreleased]

** [1.1.0] - 2022-05-17
*** Fixed
+ Spawn mpv with start time specified

** [1.0.0] - 2022-05-16
*** Changed
+ Use 'mpv:' link type instead of 'youtube:'
+ Rename to 'ol-mpv' from 'ol-youtube'
*** Added
+ Test cases for some functions
+ Github Actions for automatic testing
+ CHANGELOG file
+ example file (file:example.org)
** [0.1.0] - 2022-05-04
*** Added
+ Initial version for youtube links
+ Originally named as 'ol-youtube'


ol-mpv-test.el
(require 'cort)
(require 'ol-mpv)

;; Test for ol-mpv/convert-time
(cort-deftest-generate ol-mpv-test/-convert-time :=
	      '(((ol-mpv/convert-time "1") 1)
		((ol-mpv/convert-time "01") 1)
		((ol-mpv/convert-time "0:01") 1)
		((ol-mpv/convert-time "00:01") 1)
		((ol-mpv/convert-time "0:00:01") 1)
		((ol-mpv/convert-time "00:00:01") 1)
		((ol-mpv/convert-time "1:00") 60)
		((ol-mpv/convert-time "01:00") 60)
		((ol-mpv/convert-time "0:01:00") 60)
		((ol-mpv/convert-time "00:01:00") 60)
		((ol-mpv/convert-time "1:00:00") 360)
		((ol-mpv/convert-time "01:00:00") 360)
		))

;; Test for ol-mpv/uri/get-type
(cort-deftest-generate ol-mpv-test/uri/get-type :eq
		       '(((ol-mpv/uri/get-type nil) nil)
			 ((ol-mpv/uri/get-type "some-file.mp4") 'filepath)
			 ((ol-mpv/uri/get-type "./some-file.mp4") 'filepath)
			 ((ol-mpv/uri/get-type "../some-file.mp4") 'filepath)
			 ((ol-mpv/uri/get-type "~/some-file.mp4") 'filepath)
			 ((ol-mpv/uri/get-type "/some-file.mp4") 'filepath)
			 ((ol-mpv/uri/get-type "https://youtube.com/watch?v=FOOBAR") 'protocol))
		       )

;; Test for ol-mpv/uri/validate
(cort-deftest ol-mpv-test/uri/validate:protocol
		       '((:string= (ol-mpv/uri/validate "https://foo") "https://foo")))

(cort-deftest ol-mpv-test/uri/validate:filepath
		       '((:string= (ol-mpv/uri/validate "ol-mpv.el") (expand-file-name "ol-mpv.el"))))

(cort-deftest ol-mpv-test/uri/validate:unreachable
	      '((:cort-error 'ol-mpv/uri/unreachable (ol-mpv/uri/validate "NO_SUCH_FILE"))))

(cort-deftest ol-mpv-test/uri/validate:not-found
	      '((:cort-error 'ol-mpv/uri/not-found-error (ol-mpv/uri/validate nil))))


ol-youtube-test.el
(require 'cort)
(require 'ol-youtube)

(cort-deftest-generate ol-youtube-test/-convert-time :=
	      '(((ol-youtube/-convert-time "1") 1)
		((ol-youtube/-convert-time "01") 1)
		((ol-youtube/-convert-time "0:01") 1)
		((ol-youtube/-convert-time "00:01") 1)
		((ol-youtube/-convert-time "0:00:01") 1)
		((ol-youtube/-convert-time "00:00:01") 1)
		((ol-youtube/-convert-time "1:00") 60)
		((ol-youtube/-convert-time "01:00") 60)
		((ol-youtube/-convert-time "0:01:00") 60)
		((ol-youtube/-convert-time "00:01:00") 60)
		((ol-youtube/-convert-time "1:00:00") 360)
		((ol-youtube/-convert-time "01:00:00") 360)
		))

(cort-deftest-generate ol-youtube-test/-create-complete-url :string=
		       '(((ol-youtube/-create-complete-url "1:00" "VIDEOID") "https://www.youtube.com/watch?v=VIDEOID&t=60")
		         ((ol-youtube/-create-complete-url "1:00" "VIDEOID" t) "https://www.youtube.com/watch?v=VIDEOID")
			 )
		       )


ol-youtube.el
;;; ol-youtube.el --- Org custom link for YouTube videos -*- lexical-binding: t; -*-

;; Copyright (C) 2022 Cj.bc-sd a.k.a Cj-bc

;; Author: Cj.bc-sd a.k.a Cj-bc <cj.bc-sd@outlook.jp>
;; Created: 29 Apr 2022
;; Keywords: multimedia, outlines
;; URL: https://github.com/Cj-bc/ol-youtube
;; Package-Version: 0.1.0

;; This file is not part of GNU Emacs.

;;; Commentary:
;; org hyperlink for Youtube video timestamps.
;; It provides link type that can specify time stamp of YouTube Video.
;; For details, please refer to README.org.

;;; Code:
(require 'ol)
(require 'org)
(require 'cl-macs)
(require 'parse-time)

(org-link-set-parameters "youtube"
			 :follow #'ol-youtube/follow
			 )

;;;; Variables
(defvar ol-youtube/-sessions (make-hash-table :test 'equal)
  "Hash table of currently running mpvs.
KEY is string and represents video-uri,
 and VALUE is mpv process object that is playing video-uri's video.
")

(defcustom ol-youtube/mpv-WM-title-template "ol-youtube mpv -- {}"
  "Template of mpv windows' title. {} wil be replaced with video-uri"
  )

;;;; --- Common utilities

(defun ol-youtube/-get-video-uri (&optional pom)
  "Get video uri from buffer.
As this process requires to access the buffer, this function
should not be called so much without care.
"
  (org-entry-get pom "OL_MPV_URI" t))

(defun ol-youtube/-convert-time (timestamp)
  "Convert (HH:)MM:SS timestamp into seconds.
Return `nil' if conversion is failed.
"
  (let* ((strict (cl-case (length timestamp)
		  ;; SS, but first S is 0
		  (1 (format "00:00:0%s" timestamp))
		  ;; SS
		  (2 (format "00:00:%s" timestamp))
		  ;; MM:SS, but first M is 0
		  (4 (format "00:0%s" timestamp))
		  ;; MM:SS
		  (5 (format "00:%s" timestamp))
		  ;; HH:MM:SS, but first H is 0
		  (7 (format "0%s" timestamp))
		  ;; HH:MM:SS
		  (8 timestamp)
		  (t (error "Time stamp format is incorrect. Expected: [HH:][MM:]SS, but got %s" timestamp))))
	(parsed-time (parse-time-string strict)))
    (when parsed-time
      (+ (elt parsed-time 0)
	 (* (elt parsed-time 1) 60)
	 (* (elt parsed-time 2) 360)))))




;;;; ol-youtube/-mpv

(defun ol-youtube/-mpv-WM-title (video-uri)
  "Return WM title of mpv for given video-uri"
  (string-replace "{}" video-uri ol-youtube/mpv-WM-title-template))

(defun ol-youtube/-mpv/terminate (video-uri)
  "Cleanup jobs after mpv"
  (let ((mpv-proc (gethash video-uri ol-youtube/-sessions)))
    (if mpv-proc
	(progn (delete-process mpv-proc)
	       (remhash video-uri ol-youtube/-sessions))
      (message (format "mpv isn't runnning for %s" video-uri))))
  )

(defun ol-youtube/-mpv/setup (video-uri)
  "Launch mpv for given video-uri
This spawns one process for mpv executable that fetch video and play.

This function pushe that process object to `ol-youtube/-sessions'.

Those processes will be killed when

+ buffer is killed
+ mpv is killed by user
"
  (unless (gethash video-uri ol-youtube/-sessions)
    (let ((mpv-proc (make-process
		      :name (format "ol-youtube mpv [%s]" video-uri)
		      :buffer nil
		      :sentinel 'ol-youtube/-mpv/sentinel
		      :connection-type 'pipe
		      :command `("mpv"
				 "--no-terminal"
				 ,(format "--title=%s" (ol-youtube/-mpv-WM-title video-uri))
				 "--no-input-terminal"
				 "--input-ipc-client=fd://0"
				 video-uri
				 ))))
      (process-put mpv-proc :video-uri video-uri)
      (puthash video-uri mpv-proc ol-youtube/-sessions)
      (add-hook 'kill-buffer-hook `(lambda ()
				     (ol-youtube/-mpv/terminate ,video-uri)) 0 t)
      )))

(defun ol-youtube/-mpv/sentinel (process event)
  "Cleanup processes when process event is occured.

Currently, any event will do cleanup. This shuold be
fixed, but I'm not sure which event I should waits for.
"
  (let ((video-uri (process-get process :video-uri)))
    (ol-youtube/-mpv/terminate video-uri)
    (message "ol-youtube: mpv for %s is closed" video-uri)))

(defun ol-youtube/-mpv/change-time (connection second)
  "Send JSON IPC through the CONNECTION to set
player head to SECOND.

CONNECTION is mpv process object, and second
is the time to set in integer.
" 
  (process-send-string
   connection
   (format "%s\n" (json-encode `(("command" . ["set_property" "time-pos" ,second]))))))

;;;; --- Follow function
(defun ol-youtube/follow (link arg)
  "Control associated mpv to jump to the timestamp.
Spawn mpv if it isn't spawned"
  (let ((video-uri (ol-youtube/-get-video-uri)))
    (unless (gethash video-uri ol-youtube/-sessions)
      (message "ol-mpv: Launching mpv for [%s]...This could take some time" video-uri)
      (ol-youtube/-mpv/setup video-uri))
    (let ((mpv-proc (gethash video-uri ol-youtube/-sessions)))
      (ol-youtube/-mpv/change-time mpv-proc (ol-youtube/-convert-time link))
    )))


(provide 'ol-youtube)

;;; ol-youtube.el ends here


ol-mpv.el
;;; ol-mpv.el --- Org custom link for YouTube videos -*- lexical-binding: t; -*-

;; Copyright (C) 2022 Cj.bc-sd a.k.a Cj-bc

;; Author: Cj.bc-sd a.k.a Cj-bc <cj.bc-sd@outlook.jp>
;; Created: 29 Apr 2022
;; Keywords: multimedia, outlines
;; URL: https://github.com/Cj-bc/ol-mpv
;; Package-Version: 1.1.0

;; This file is not part of GNU Emacs.

;;; Commentary:
;; org hyperlink for Youtube video timestamps.
;; It provides link type that can specify time stamp of YouTube Video.
;; For details, please refer to README.org.

;;; Code:
(require 'ol)
(require 'org)
(require 'cl-macs)
(require 'parse-time)

(org-link-set-parameters "mpv"
			 :follow #'ol-mpv/follow
			 )

;;;; Variables
(defvar ol-mpv/sessions (make-hash-table :test 'equal)
  "Hash table of currently running mpvs.
KEY is string and represents video-uri,
 and VALUE is mpv process object that is playing video-uri's video.
")

(defgroup ol-mpv '((ol-mpv/mpv-WM-title-template custom-variable))
  "Group for ol-mpv library"
  :group 'external
  :group 'org
  )
(defcustom ol-mpv/mpv-WM-title-template "ol-mpv mpv -- {}"
  "Template of mpv windows' title. {} wil be replaced with video-uri"
  :type (string)
  )


;; Symbol that indicates URI type.
;;;; --- Common utilities

(defun ol-mpv/get-video-uri (&optional pom)
  "Get video uri from buffer.
As this process requires to access the buffer, this function
should not be called so much without care.
"
  (org-entry-get pom "OL_MPV_URI" t))

(defun ol-mpv/convert-time (timestamp)
  "Convert (HH:)MM:SS timestamp into seconds.
Return `nil' if conversion is failed.
"
  (let* ((strict (cl-case (length timestamp)
		  ;; SS, but first S is 0
		  (1 (format "00:00:0%s" timestamp))
		  ;; SS
		  (2 (format "00:00:%s" timestamp))
		  ;; MM:SS, but first M is 0
		  (4 (format "00:0%s" timestamp))
		  ;; MM:SS
		  (5 (format "00:%s" timestamp))
		  ;; HH:MM:SS, but first H is 0
		  (7 (format "0%s" timestamp))
		  ;; HH:MM:SS
		  (8 timestamp)
		  (t (error "Time stamp format is incorrect. Expected: [HH:][MM:]SS, but got %s" timestamp))))
	(parsed-time (parse-time-string strict)))
    (when parsed-time
      (+ (elt parsed-time 0)
	 (* (elt parsed-time 1) 60)
	 (* (elt parsed-time 2) 360)))))




;;;; ol-mpv/mpv

(defun ol-mpv/mpv-WM-title (video-uri)
  "Return WM title of mpv for given video-uri"
  (string-replace "{}" video-uri ol-mpv/mpv-WM-title-template))

(defun ol-mpv/mpv/terminate (video-uri)
  "Cleanup jobs after mpv"
  (let ((mpv-proc (gethash video-uri ol-mpv/sessions)))
    (if mpv-proc
	(progn (delete-process mpv-proc)
	       (remhash video-uri ol-mpv/sessions))
      (message (format "mpv isn't runnning for %s" video-uri))))
  )

(defun ol-mpv/mpv/setup (video-uri &optional initial-pos)
  "Launch mpv for given video-uri
This spawns one process for mpv executable that fetch video and play.

This function pushe that process object to `ol-mpv/sessions'.

Those processes will be killed when

+ buffer is killed
+ mpv is killed by user
"
  (unless (gethash video-uri ol-mpv/sessions)
    (let ((mpv-proc (make-process
		      :name (format "ol-mpv mpv [%s]" video-uri)
		      :buffer nil
		      :sentinel 'ol-mpv/mpv/sentinel
		      :connection-type 'pipe
		      :command `("mpv"
				 "--no-terminal"
				 ,(format "--title=%s" (ol-mpv/mpv-WM-title video-uri))
				 ,(format "--start=%s" (or initial-pos 0))
				 "--no-input-terminal"
				 "--input-ipc-client=fd://0"
				 ,video-uri
				 ))))
      (process-put mpv-proc :video-uri video-uri)
      (puthash video-uri mpv-proc ol-mpv/sessions)
      (add-hook 'kill-buffer-hook `(lambda ()
				     (ol-mpv/mpv/terminate ,video-uri)) 0 t)
      )))

(defun ol-mpv/mpv/sentinel (process _)
  "Cleanup processes when process event is occured.

Currently, any event will do cleanup. This shuold be
fixed, but I'm not sure which event I should waits for.
"
  (let ((video-uri (process-get process :video-uri)))
    (ol-mpv/mpv/terminate video-uri)
    (message "ol-mpv: mpv for %s is closed" video-uri)))

(defun ol-mpv/mpv/change-time (connection second)
  "Send JSON IPC through the CONNECTION to set
player head to SECOND.

CONNECTION is mpv process object, and second
is the time to set in integer.
" 
  (process-send-string
   connection
   (format "%s\n" (json-encode `(("command" . ["set_property" "time-pos" ,second]))))))


;;;; --- ol-mpv/uri
(defun ol-mpv/uri/validate (uri)
  "Return validated URI or throw error signal.

This only ensure that:
+ it is non-nil
+ if it's local file path, it exists

Possible errors:
+ `ol-mpv/uri/not-found-error', when given URI is `nil'
+ `ol-mpv/uri/unreachable', when given URI is local filepath & not found
+ Raw error when something wrong happend
"
  (unless uri
    (signal 'ol-mpv/uri/not-found-error nil))
  (pcase (ol-mpv/uri/get-type uri)
    ('filepath (if (file-exists-p uri)
		  (expand-file-name uri)
		(signal 'ol-mpv/uri/unreachable nil)))
    ('protocol uri)
    (_ (error "ol-mpv: unknown error occured"))
    ))

(defun ol-mpv/uri/get-type (uri)
  "Return symbol of URI type

Possible symbols are:

+ 'protocol for URI that start with \"<protocol>://\"
+ 'filepath for other path
+ nil for nil
"
  (when uri
      (if (string-match-p "://" uri)
	  'protocol
	'filepath
	  )))

;;;; --- Errors
(define-error 'ol-mpv-error "ol-mpv: Generic error for ol-mpv packge")
(define-error 'ol-mpv/uri/not-found-error "ol-mpv: Couldn't find URI" 'ol-mpv-error)
(define-error 'ol-mpv/uri/unreachable "ol-mpv: Couldn't reach URI" 'ol-mpv-error)

;;;; --- Follow function
(defun ol-mpv/follow (link _)
  "Control associated mpv to jump to the timestamp.
Spawn mpv if it isn't spawned"
  (let ((video-uri (ol-mpv/uri/validate (ol-mpv/get-video-uri))))
    (pcase (gethash video-uri ol-mpv/sessions)
      ('nil (message "ol-mpv: Launching mpv for [%s]...This could take some time" video-uri)
	    (ol-mpv/mpv/setup video-uri (ol-mpv/convert-time link)))
      (mpv-proc (ol-mpv/mpv/change-time mpv-proc (ol-mpv/convert-time link))))))


(provide 'ol-mpv)

;;; ol-mpv.el ends here
